src/components/heatmap/heatmap.tsx

```diff
@@ -0,0 +1,344 @@
+   */
+
+  const addData = (tuple: any) => {
+    const configData = Object.assign({}, config.data);
+    let hm = null;
+    const x = tuple.data[configData.xColumn];
+    const y = tuple.data[configData.yColumn];
+    const z = tuple.data[configData.zColumn];
+    const title = tuple.data[configData.titleColumn];
+    const id = tuple.data[configData.idColumn];
+    for (let i = 0; i < heatmaps.length; i++) {
+      if (Object(heatmaps[i]).title === title) {
+        Object(heatmaps)[i]['id'] = id;
+        hm = heatmaps[i];
+      }
+    }
+    if (hm === null) {
+      hm = { 'title': title, 'rows': { y: [], x: [], z: [], type: 'heatmap' } };
+      heatmaps.push(hm);
+    }
+    let rowIndex = Object(hm).rows.y.indexOf(y);
+    if (rowIndex < 0) {
+      Object(hm).rows.y.push(y);
+      Object(hm).rows.z.push([]);
+      rowIndex = Object(hm).rows.y.indexOf(y);
+    }
+    if (!Object(hm).rows.x.includes(x)) {
+      Object(hm).rows.x.push(x);
+    }
+    Object(hm).rows.z[rowIndex].push(z);
+  }
+
+  /**
+     * @param {object} input : Input parameters of heatmap directive
+     * @param {number} longestXTick : Length of longest X axis label
+     * @param {number} longestYTick : Length of longest Y axis label
+     * @param {number} lengthY : Number of Y values
+     * Calculates the height and margins of the heatmap based on the number of y values and length of the longest X label
+     * so that the labels do not get clipped and the bar height is adjusted accordingly.
+     * Return an object with all the required layout parameters.
+     * @example
+     * {
+     * 	height: height of the heatmap,
+     * 	width: width of the heatmap,
+     * 	margin: {
+     * 		t: top margin of the heatmap,
+     * 		r: right margin of the heatmap,
+     * 		b: bottom margin of the heatmap,
+     * 		l: left of the heatmap
+     * 	},
+     * 	xTickAngle: inclination of x axis labels,
+     *  yTickAngle: inclination of y axis labels,
+     * 	tickFont: font to be used in labels
+     * }
+     */
+  const getHeatmapLayoutParams = (input: any, longestXTick: number, longestYTick: number, lengthY: number) => {
+    let height;
+    let yTickAngle;
+    const tMargin = 25;
+    let rMargin, bMargin, lMargin;
+
+    if (longestXTick <= 18) {
+      height = longestXTick * 9 + lengthY * 10 + 50;
+      bMargin = 8.4 * longestXTick;
+    } else if (longestXTick <= 22) {
+      height = longestXTick * 9 + lengthY * 10 + 55;
+      bMargin = 8.4 * longestXTick;
+    } else if (longestXTick <= 30) {
+      height = longestXTick * 8.8 + lengthY * 10 + 55;
+      bMargin = 8.2 * longestXTick;
+    } else {
+      height = longestXTick * 8.8 + lengthY * 10 + 45;
+      bMargin = 8 * longestXTick;
+    }
+
+    if (lengthY === 1) {
+      yTickAngle = -90;
+      lMargin = 30;
+      rMargin = 20;
+    } else {
+      yTickAngle = 0;
+      lMargin = 20 + longestYTick * 7;
+      rMargin = 5;
+    }
+
+    const layoutParams = {
```

Comment:
This should have a type as well. Similar comment to the one I made below about `inputParams`.